{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Wo0Vq3ZvVPOq",
        "outputId": "19fdf542-4692-4af8-9497-8d6cf4130830"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: plotly in /usr/local/lib/python3.10/dist-packages (5.15.0)\n",
            "Requirement already satisfied: tenacity>=6.2.0 in /usr/local/lib/python3.10/dist-packages (from plotly) (8.3.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from plotly) (24.0)\n",
            "Collecting BioPython\n",
            "  Downloading biopython-1.83-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m8.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from BioPython) (1.25.2)\n",
            "Installing collected packages: BioPython\n",
            "Successfully installed BioPython-1.83\n",
            "Collecting Bio\n",
            "  Downloading bio-1.7.0-py3-none-any.whl (279 kB)\n",
            "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m279.2/279.2 kB\u001b[0m \u001b[31m5.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hRequirement already satisfied: biopython>=1.80 in /usr/local/lib/python3.10/dist-packages (from Bio) (1.83)\n",
            "Collecting gprofiler-official (from Bio)\n",
            "  Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
            "Collecting mygene (from Bio)\n",
            "  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.10/dist-packages (from Bio) (2.0.3)\n",
            "Requirement already satisfied: pooch in /usr/local/lib/python3.10/dist-packages (from Bio) (1.8.1)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.10/dist-packages (from Bio) (2.31.0)\n",
            "Requirement already satisfied: tqdm in /usr/local/lib/python3.10/dist-packages (from Bio) (4.66.4)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython>=1.80->Bio) (1.25.2)\n",
            "Collecting biothings-client>=0.2.6 (from mygene->Bio)\n",
            "  Downloading biothings_client-0.3.1-py2.py3-none-any.whl (29 kB)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2023.4)\n",
            "Requirement already satisfied: tzdata>=2022.1 in /usr/local/lib/python3.10/dist-packages (from pandas->Bio) (2024.1)\n",
            "Requirement already satisfied: platformdirs>=2.5.0 in /usr/local/lib/python3.10/dist-packages (from pooch->Bio) (4.2.1)\n",
            "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from pooch->Bio) (24.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (3.3.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (3.7)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (2.0.7)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests->Bio) (2024.2.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.10/dist-packages (from python-dateutil>=2.8.2->pandas->Bio) (1.16.0)\n",
            "Installing collected packages: gprofiler-official, biothings-client, mygene, Bio\n",
            "Successfully installed Bio-1.7.0 biothings-client-0.3.1 gprofiler-official-1.0.0 mygene-3.2.2\n"
          ]
        }
      ],
      "source": [
        "!pip install plotly\n",
        "!pip install BioPython\n",
        "!pip install Bio\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import matplotlib.pyplot as plt\n",
        "import IPython.display as ipd\n",
        "import plotly.graph_objects as go\n",
        "from Bio.Seq import Seq\n",
        "from Bio import Entrez, SeqIO\n",
        "from Bio.SeqUtils import molecular_weight\n",
        "from Bio.SeqUtils import MeltingTemp as mt\n",
        "\n",
        "\n",
        "\n",
        "class DNASequence:\n",
        "    def __init__(self, sequence, label='DNA Sequence'):\n",
        "        \"\"\"Function initializes the DNASequence object with sequence and label\"\"\"\n",
        "        self.sequence = sequence.upper() #Convert sequence to uppercase\n",
        "        self.label = label #Label for the sequence\n",
        "        self.start_codons = ['ATG', 'TTG', 'CTG', 'GTG'] #List of start codons\n",
        "        self.stop_codons = ['TAA', 'TAG', 'TGA'] #List of stop codons\n",
        "\n",
        "\n",
        "    def get_gc_content(self, subseq=None):\n",
        "      ''' Function computes GC content of input sequence'''\n",
        "      sequence = subseq if subseq is not None else self.sequence  #Use subsequence if provided\n",
        "      gc_count = sum(1 for base in sequence if base in \"GC\") #Count G and C bases\n",
        "      total_bases = sum(1 for base in sequence if base in 'ATCG') #Count all valid base\n",
        "      return (gc_count / total_bases) * 100 if total_bases > 0 else 0 #Calculate GC content percentage\n",
        "\n",
        "\n",
        "    def orf_seqs(self):\n",
        "        ''' Function computes list of open reading frames of input sequence'''\n",
        "        orfs = [] #Initialize list to store ORFs\n",
        "        for i in range(len(self.sequence)):\n",
        "          if self.sequence[i:i+3] in self.start_codons: #Check for start codon\n",
        "            for j in range(i+3, len(self.sequence), 3):\n",
        "              if self.sequence[j:j+3] in self.stop_codons: #Check for stop codon\n",
        "                 orfs.append(self.sequence[i:j+3]) #Append ORF to list\n",
        "        return orfs\n",
        "\n",
        "    def transcription(self):\n",
        "        ''' Function computes the transcript of input sequence'''\n",
        "        complement = {'A': 'U', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'} #DNA to RNA complement\n",
        "        return ''.join(complement.get(base, 'N') for base in self.sequence) #Transcribe to RNA\n",
        "\n",
        "    def id_palindromes(self):\n",
        "        \"\"\"Function identifies palindromic sequences in the DNA sequence\"\"\"\n",
        "        pal = [] #Initialize list to store palindromes\n",
        "        for i in range(len(self.sequence) - 1):\n",
        "         for j in range(i+4, len(self.sequence)):\n",
        "          if self.sequence[i:j] == self.sequence[j-1:i-1:-1]: #Check if sequence is palindrome\n",
        "            pal.append(self.sequence[i:j])\n",
        "        return pal\n",
        "\n",
        "    def count_kmers(self, k):\n",
        "      ''' Function computes kmers of specified length of input sequence'''\n",
        "      kmers = {} #Initialize dictionary to store k-mer counts\n",
        "      for i in range(len(self.sequence) - k + 1):\n",
        "        kmer = self.sequence[i:i+k] #Extract k-mer\n",
        "        kmers[kmer] = kmers.get(kmer, 0) + 1 #Count k-mer occurrences\n",
        "      return kmers\n",
        "\n",
        "    def melting_temperature(self):\n",
        "        ''' Function computes melting temperature of input sequence'''\n",
        "        gc_cont = (self.sequence.count(\"G\") + self.sequence.count(\"C\"))/len(self.sequence) #Calculate GC content\n",
        "        melt_temp = (64.9 + 41 * (gc_cont - 16.4) )/len(self.sequence) #Calculate melting temperature\n",
        "        return melt_temp\n",
        "\n",
        "\n",
        "    def design_primers(self, length=20, min_tm=52, max_tm=58, min_gc=40, max_gc=60):\n",
        "      ''' Function returns a list of primers of length 20 from the input sequence'''\n",
        "      init_primers = [] #Initialize list to store primer candidates\n",
        "      for i in range(len(self.sequence) - length + 1):\n",
        "       seq = self.sequence[i:i+length] #Extract potential primer\n",
        "       gc_content = self.get_gc_content(seq) #Calculate GC content\n",
        "       tm = mt.Tm_Wallace(seq) #Calculate melting temperature\n",
        "       if min_tm <= tm <= max_tm and min_gc <= gc_content <= max_gc: #Check constraints\n",
        "          init_primers.append(seq) #Append valid primer to list\n",
        "       return init_primers\n",
        "\n",
        "    def has_hairpin(self,seq):\n",
        "      ''' Function check if there are haiprin structures in input sequence'''\n",
        "      comp = 0\n",
        "      not_comp = 0\n",
        "      for i in range(10):\n",
        "        to_check_seq_1 = seq[i] #Base at position i from the start of the sequence\n",
        "        to_check_seq_2 = seq[-i-1] #Base at position i from the end of the sequence\n",
        "        if to_check_seq_1 == ComplementarySequence(to_check_seq_2).get_reverse_complement():\n",
        "          if not_comp > 0:\n",
        "            return False\n",
        "          comp += 1  #Pair\n",
        "        else:\n",
        "            not_comp += 1\n",
        "      if comp > 0 and 2 <= not_comp <= 4:\n",
        "       return True\n",
        "      return False\n",
        "\n",
        "    def filter_primers(self, primers):\n",
        "      ''' Function filters out primers without any hairpin '''\n",
        "      return [p for p in primers if not self.has_hairpin(p)]\n",
        "\n",
        "    def save_results(self):\n",
        "      ''' Function opens a single file and writes all the results to it'''\n",
        "      #Open a single file to write all results\n",
        "      with open(f\"{self.label}_all_results.txt\", 'w') as f:\n",
        "        f.write(f'Accession Number: {self.label}\\n')\n",
        "        f.write(f\"GC%: {self.get_gc_content():.2f}\\n\")\n",
        "        f.write(f\"Transcript: {self.transcription()}\\n\")\n",
        "\n",
        "        #Complementary and Reverse Complementary Sequences\n",
        "        comp_seq = ComplementarySequence(self.sequence).get_complementary_sequence()\n",
        "        rev_comp_seq = ComplementarySequence(self.sequence).get_reverse_complement()\n",
        "        f.write(f\"Double stranded DNA\\n\")\n",
        "        f.write(f\"5' {self.sequence} 3' \\n\")\n",
        "        f.write(f\" {''.join(['|'for c in range(len(self.sequence))])}\\n\")\n",
        "        f.write(f\"3' {comp_seq} 5' \\n\")\n",
        "        f.write(f'Reverse Complementary Sequence: {rev_comp_seq}\\n\\n')\n",
        "        palindromes = self.id_palindromes()\n",
        "        f.write(f\"Palindromic Sequences in DNA sequence: {palindromes}\\n\")\n",
        "\n",
        "        #Melting Temperature of DNA\n",
        "        tm = self.melting_temperature()\n",
        "        f.write(f\"Melting Temperature:{tm:.2f} °C \\n\")\n",
        "\n",
        "\n",
        "        #Protein Translations\n",
        "        translation = Translation(self.sequence)\n",
        "        frames = translation.translate_all_frames()\n",
        "        f.write('Protein Translation Frames:\\n')\n",
        "\n",
        "        #Protein Weights\n",
        "        annot_prot = translation.annotate_protein()\n",
        "        f.write(f\"Molecular Weights of Proteins:{annot_prot}\\n\")\n",
        "\n",
        "        for f_name, prot in frames.items():\n",
        "          f.write(f\"{f_name}: {prot}\\n\")\n",
        "          f.write(\"\\n\")\n",
        "        #ORF Positions and Sequences\n",
        "        orf_positions = self.orf_pos_finder()\n",
        "        orf_sequences = self.orf_seqs()\n",
        "        f.write(\"ORF Positions and Sequences:\\n\")\n",
        "        for p, seq in zip(orf_positions, orf_sequences):\n",
        "          f.write(f\"Position: {p}, Sequence: {seq}\\n\")\n",
        "        f.write(\"\\n\")\n",
        "\n",
        "        #Primer Design\n",
        "        primers = self.design_primers()\n",
        "        f_primers = self.filter_primers(primers)\n",
        "        f.write(\"Initial Primer Candidates:\\n\")\n",
        "        for x in primers:\n",
        "          f.write(f\"{x}\\n\")\n",
        "        f.write(\"Filtered Primers:\\n\")\n",
        "        for y in f_primers:\n",
        "         f.write(f\"{y}\\n\")\n",
        "        f.write(\"\\n\")\n",
        "\n",
        "        #K-mer Count\n",
        "        k = 3\n",
        "        k_mers = self.count_kmers(k)\n",
        "        f.write(\"K-mer Count:\\n\")\n",
        "        for kmer, count in k_mers.items():\n",
        "          f.write(f\"{kmer}: {count}\\n\")\n",
        "\n",
        "    def plot_kmer_distribution(self, k=3):\n",
        "      ''' Function creates a barplot to display kmer '''\n",
        "      k = 3\n",
        "      k_mers = self.count_kmers(k) #Count the frequency of each k-mer in the sequence\n",
        "      kdf = pd.DataFrame(list(k_mers.items()), columns=['Kmer', 'Frequency'])\n",
        "      plt.figure(figsize=(10, 6))\n",
        "      sns.barplot(x=\"Kmer\", y=\"Frequency\", data=kdf)\n",
        "      plt.title(f'{k}mer Frequency Distribution')\n",
        "      plt.xticks(rotation=90)\n",
        "      plt.show()\n",
        "\n",
        "\n",
        "\n",
        "    def orf_pos_finder(self):\n",
        "        ''' Function checks for position of open reading frame of the input sequence '''\n",
        "        orfs = []\n",
        "        i = 0\n",
        "        while i < len(self.sequence) - 2:\n",
        "          if self.sequence[i:i+3] in self.start_codons: #Record start position\n",
        "            start = i #Record start position\n",
        "            i += 3\n",
        "            for j in range(i , len(self.sequence)-2, 3 ):\n",
        "              if self.sequence[j:j+3] in self.stop_codons: #Check for stop codon\n",
        "                end = j + 2 #Record end position\n",
        "                if end - start > 100: #Check ORF length\n",
        "                  orfs.append((start, end)) #Append ORF position\n",
        "                  break\n",
        "          else:\n",
        "            i += 3\n",
        "        return orfs\n",
        "\n",
        "    def orf_visualizer(self):\n",
        "      ''' Function creates an interactive open reading frame visualizer'''\n",
        "      orfs_pos = self.orf_pos_finder() #Find positions of ORFs in the sequence\n",
        "      fig = go.Figure() #Create a new figure for visualization\n",
        "\n",
        "      for idx, (start, end) in enumerate(orfs_pos, 1): #Iterate over each ORF position to add it to the figure\n",
        "        orf_length = end - start\n",
        "        if orf_length > 100: #Only add ORFs longer than 100 bases to the figure\n",
        "          fig.add_trace(go.Scatter(x=[start, end], y=[1, 1], mode='lines+markers+text', line=dict(color='tomato', width=4),\n",
        "            text=f\"ORF{idx}\", textposition=\"bottom center\", hoverinfo='text',\n",
        "            hovertext=f\"<b>ORF{idx}</b><br>Start: {start}<br>End: {end}<br>Length: {orf_length} bases\",\n",
        "            marker=dict(color='tomato', size=10, opacity=0), showlegend=False))\n",
        "\n",
        "          #Update layout to improve visualization\n",
        "      fig.update_layout(title='Visualization of ORFs', xaxis_title='Position in Sequence', yaxis_title='Track',\n",
        "                      yaxis=dict(showticklabels=False, range=[0.8, 1.2]),  #Adjust y-axis to prevent cutting off text\n",
        "        plot_bgcolor='white', xaxis=dict(showgrid=True, gridcolor='LightGrey', zeroline=False, rangeslider=dict(visible=True),type='linear'))\n",
        "      fig.update_layout(width=1200, height=400) #Set figure dimensions\n",
        "      fig.show()   #Display the figure\n",
        "\n",
        "\n",
        "class ComplementarySequence(DNASequence):\n",
        "    def get_complementary_sequence(self):\n",
        "        ''' Function returns the complementary sequence of input sequence '''\n",
        "        #Mapping for standard bases plus 'N'\n",
        "        complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C', 'N': 'N'}\n",
        "        #Default to 'N' if unexpected character\n",
        "        return ''.join(complement.get(base, 'N') for base in self.sequence)\n",
        "\n",
        "    def get_reverse_complement(self):\n",
        "      ''' Function returns the reverse complement of the input sequence '''\n",
        "      return self.get_complementary_sequence()[::-1]\n",
        "\n",
        "\n",
        "class Translation(DNASequence):\n",
        "    def __init__(self, sequence):\n",
        "        \"\"\"Function initializes the Translation object with the input sequence and codon table.\"\"\"\n",
        "        super().__init__(sequence)\n",
        "        self.codon_table = {\"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
        "            \"TGT\": \"C\", \"TGC\": \"C\",\"GAT\": \"D\", \"GAC\": \"D\",\"GAA\": \"E\", \"GAG\": \"E\",\n",
        "            \"TTT\": \"F\", \"TTC\": \"F\",\"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
        "            \"CAT\": \"H\", \"CAC\": \"H\",\"ATA\": \"I\", \"ATT\": \"I\", \"ATC\": \"I\",\n",
        "            \"AAA\": \"K\", \"AAG\": \"K\",\"TTA\": \"L\", \"TTG\": \"L\", \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
        "            \"ATG\": \"M\",\"AAT\": \"N\", \"AAC\": \"N\",\"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
        "            \"CAA\": \"Q\", \"CAG\": \"Q\",\"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\", \"AGA\": \"R\", \"AGG\": \"R\",\n",
        "            \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\", \"AGT\": \"S\", \"AGC\": \"S\",\n",
        "            \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
        "            \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
        "            \"TGG\": \"W\",\"TAT\": \"Y\", \"TAC\": \"Y\", \"TAA\": \"_\", \"TAG\": \"_\", \"TGA\": \"_\"}\n",
        "\n",
        "    def translate_frame(self, frame):\n",
        "        ''' Function translates the input sequence '''\n",
        "        protein_sequence = Seq(self.sequence[frame:]).translate(to_stop=True) #Translate the sequence taking care of the frame shift\n",
        "        return str(protein_sequence)  #Convert the protein sequence to a string and return it\n",
        "\n",
        "\n",
        "    def translate_all_frames(self):\n",
        "        ''' Function translates all the frames of given sequence'''\n",
        "        frames = {}\n",
        "        #3 forward frames\n",
        "        for frame in range(3):\n",
        "          frames[f\"Forward Frame {frame+1}\"] = self.translate_frame(frame)\n",
        "        #3 reverse frames\n",
        "        reverse_seq = ComplementarySequence(self.sequence).get_reverse_complement()\n",
        "        for frame in range(3):\n",
        "          frames[f\"Reverse Frame {frame+1}\"] = Translation(reverse_seq).translate_frame(frame)\n",
        "        return frames\n",
        "\n",
        "    def calc_mol_weight(self, protein_sequence):\n",
        "      ''' Function computes the moleecular weight of input protein'''\n",
        "      #Calculate the molecular weight using BioPython\n",
        "      return molecular_weight(protein_sequence, seq_type='protein')\n",
        "\n",
        "    def annotate_protein(self):\n",
        "      ''' Function returns a dictionary of molecular weights of proteins translated from input sequence '''\n",
        "      all_frames = self.translate_all_frames()\n",
        "      annotated_frames = {}\n",
        "      for frame, seq in all_frames.items():\n",
        "          if seq:  #Ensure sequence is not empty\n",
        "            mw = self.calc_mol_weight(seq)\n",
        "            annotated_frames[frame] = {'Sequence': seq, 'Molecular Weight': mw}\n",
        "      return annotated_frames\n",
        "\n",
        "\n",
        "class EntrezData:\n",
        "    def __init__(self, email):\n",
        "        Entrez.email = email\n",
        "\n",
        "    def fetch_sequence(self, accession):\n",
        "      ''' Function fetches fasta sequence associated with accesion id '''\n",
        "      with Entrez.efetch(db=\"nucleotide\", rettype=\"fasta\", retmode=\"text\", id=accession) as handle:\n",
        "          #Read the FASTA record\n",
        "          seq_record = SeqIO.read(handle, \"fasta\")\n",
        "          #Extract the sequence from the record\n",
        "          sequence = str(seq_record.seq)\n",
        "          #print(\"Sequence obtained from Entrez:\", sequence)\n",
        "          print(\"Length of sequence obtained from Entrez:\", len(sequence))\n",
        "      return DNASequence(sequence)\n",
        "\n",
        "\n",
        "def main():\n",
        "    ''' Main function that computes everything '''\n",
        "    #Prompt the user to enter their email address for Entrez\n",
        "    email = input(\"Please enter your email address (for Entrez): \")\n",
        "    entrez_data = EntrezData(email)\n",
        "\n",
        "    #Prompt the user to enter accession numbers separated by commas\n",
        "    accession_input = input(\"Please enter accession numbers separated by commas: \")\n",
        "    accession_numbers = [acc.strip() for acc in accession_input.split(',')]\n",
        "\n",
        "    #Loop through each accession number to fetch sequences and process them\n",
        "    for index, accession in enumerate(accession_numbers):\n",
        "        try:\n",
        "            sequence = entrez_data.fetch_sequence(accession) #Fetch the DNA sequence using the accession number\n",
        "            sequence.label = accession  #Label each sequence by its accession number\n",
        "            sequence.save_results() #Save the results to a file\n",
        "            print(f\"Results saved for accession {accession}\")\n",
        "\n",
        "            #Visual separator for console output\n",
        "            print(\"#\" * 20 + \" Next Accession Number \" + \"#\" * 20 + \"\\n\")\n",
        "\n",
        "            #Plot the k-mer distribution for the sequence\n",
        "            sequence.plot_kmer_distribution()\n",
        "            print(\"###############################################################\")\n",
        "\n",
        "            #Visualize the open reading frames (ORFs) for the sequence\n",
        "            sequence.orf_visualizer()\n",
        "\n",
        "\n",
        "            #Visual separator for console output if there are more accession numbers to process\n",
        "            if index < len(accession_numbers) - 1:\n",
        "                print(\"#\" * 20 + \" Next Accession Number \" + \"#\" * 20 + \"\\n\")\n",
        "        except Exception as e:\n",
        "            #Handle any errors that occur during sequence fetching and processing\n",
        "            print(f\"Error in fetching sequence for accession id number {accession}: {e}\")\n",
        "\n",
        "\n",
        "if __name__ == '__main__':\n",
        "    main()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 332
        },
        "id": "Pivii4S1VoM3",
        "outputId": "e868384e-aec1-4a9d-9bb7-b07401c87980"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "Interrupted by user",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-10-10ac7ed32a09>\u001b[0m in \u001b[0;36m<cell line: 324>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    323\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    324\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'__main__'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 325\u001b[0;31m     \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m<ipython-input-10-10ac7ed32a09>\u001b[0m in \u001b[0;36mmain\u001b[0;34m()\u001b[0m\n\u001b[1;32m    294\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    295\u001b[0m     \u001b[0;34m''' Main function that computes everything '''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 296\u001b[0;31m     \u001b[0memail\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Please enter your email address (for Entrez): \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    297\u001b[0m     \u001b[0mentrez_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mEntrezData\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0memail\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    298\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36mraw_input\u001b[0;34m(self, prompt)\u001b[0m\n\u001b[1;32m    849\u001b[0m                 \u001b[0;34m\"raw_input was called, but this frontend does not support input requests.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    850\u001b[0m             )\n\u001b[0;32m--> 851\u001b[0;31m         return self._input_request(str(prompt),\n\u001b[0m\u001b[1;32m    852\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_ident\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    853\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_parent_header\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py\u001b[0m in \u001b[0;36m_input_request\u001b[0;34m(self, prompt, ident, parent, password)\u001b[0m\n\u001b[1;32m    893\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    894\u001b[0m                 \u001b[0;31m# re-raise KeyboardInterrupt, to truncate traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 895\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyboardInterrupt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Interrupted by user\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    896\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwarning\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid Message:\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc_info\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: Interrupted by user"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "n_t_KWTXpx-8"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}